sampling : add XTC sampler

MaggotHATE opened the pull request:
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

This is my implementation of the XTC sampler as described by @p-e-w in https://github.com/oobabooga/text-generation-webui/pull/6335 , the main difference is `threshold_max` parameter which seems to help with finetunes. Additional "thank you" goes to @LostRuins who implemented XTC a bit earlier than me (probabilities + logits system is a bit confusing at first). 

> XTC is a novel sampler that turns truncation on its head: Instead of pruning the least likely tokens, under certain circumstances, it removes the most likely tokens from consideration.

Technically, this sampler: 
* checks for probability `xtc_p`
* searches for all tokens with probabilities above threshold `xtc_t` and below upper threshold `xtc_t_max`
* penalizes and removes all such tokens besides the one least probable of them.

This means that at least one "most probable" token is always left untouched, which should keep things stable. For more details and explanations it is better to read @p-e-w 's original PR as it is very thorough and comprehensive. 

Additionally, I have not encountered problems with `eos` tokens that were described in discussions of the original implementation, so I have not added any checks for it.

Being a unique sampler, XTC is not included in the queue by default (like Mirostats)

In my experience, the effect varies between models and even sizes of models. XTC improves creativity greatly, but may break models in specific cases (for example, multi-language capabilities of Nemo can be slightly worsened for inflective languages). With larger models XTC may require very careful choice of parameters - I have seen more refusals in Mistral Large 123b with default XTC settings, for example (although, not on my implementation, but through Kobold.cpp).

In general, however, XTC is a really interesting idea and a sampler worth testing.

MaggotHATE committed:
Initial XTC commit

Adds XTC sampler, not activated by default, but recommended settings by default.

MaggotHATE committed:
Cleanup

MaggotHATE committed:
Simplified chances calculation

To be more inline with the original implementation, chance is calculated once at the beginning.

pnb commented on src/llama-sampling.cpp:
```diff
@@ -1059,6 +1059,86 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
+        std::sort(cur_p->data, cur_p->data + cur_p->size, [](const llama_token_data & a, const llama_token_data & b) {
+            return a.logit > b.logit;
+        });
+        cur_p->sorted = true;
```
Another dumb question for my own edification, but doesn't the `for` loop above need to assume the logits are already sorted? And so wouldn't `cur_p->sorted` already be true?

pnb reviewed (commented):
I'm not very familiar with the code base, but XTC is a very interesting idea to me so I took a read through to see how you implemented it, and left a few comments in case they are helpful. In case not, feel free to ignore. :)

MaggotHATE commented on common/arg.cpp:
```diff
@@ -966,6 +966,27 @@ gpt_params_context gpt_params_parser_init(gpt_params & params, llama_example ex,
+    ).set_sparam());
+    add_opt(llama_arg(
+        {"--xtc-t"}, "N",
+        format("xtc threshold (default: %.1f, 0.0 = disabled)", (double)params.sparams.xtc_t),
```
It technically is correct, and both ways, I will fix it both in description and in the sampler itself. It doesn't make sense both as 0 (all tokens affected) and 1 (no tokens affected).

p-e-w replied:
Setting `threshold` to 0 should not disable XTC. It should eliminate all tokens except for the least probable one, in line with the XTC mechanism, since all tokens have probability >= 0. Anything else is an unnecessary and confusing special case, given that setting `probability` to 0 already disables XTC in a natural and intuitive manner.

And there are cases where it *does* make sense to do this. In fact, it's quite an interesting option to have when combining XTC with aggressive truncation.

MaggotHATE replied:
Yes, sorry, I've missed this comment somehow, will fix.

MaggotHATE commented on common/common.h:
```diff
@@ -108,6 +109,9 @@ struct gpt_sampler_params {
     int32_t top_k             = 40;    // <= 0 to use vocab size
     float   top_p             = 0.95f; // 1.0 = disabled
     float   min_p             = 0.05f; // 0.0 = disabled
+    float   xtc_p             = 0.50f; // 0.0 = disabled
```
As I've mentioned, it is excluded from the samplers sequence by default, like Mirostat.

pnb replied:
Ah right, makes sense!

MaggotHATE commented on src/llama-sampling.cpp:
```diff
@@ -1059,6 +1059,86 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
+
+    int removed = 0;
+    // going through all candidates from back to front, easier to keep the last of probables
+    for (int i = (cur_p->size - 1); i >= 0; --i) {
```
I kept this version since XTC is not meant to be used before truncating samplers. The recommended sequence is "min-p" -> "xtc", so the amount of candidates should not be too long. In practice, I have not noticed it affecting tg speed even with no prior truncation. Plus, this version keeps the code a bit cleaner.

As for sorting, it is the same as in `llama_sampler_softmax_impl`. Your idea is good though, I will look into that.

p-e-w replied:
I'm quite sure performance will be visibly affected if there is a high baseline speed and a large vocabulary (some models have 200k tokens in their vocabulary nowadays). If you are seeing 80 t/s normally, then there's no way you can loop over 200k candidates 80 times per second on most hardware, in addition to all the other TG work.

The TGWUI implementation uses PyTorch vector operations for high performance; maybe something equivalent can be done here.

MaggotHATE replied:
> some models have 200k tokens

Normally that amount of token will be pruned first by `min_p`: in my tests it narrowed down to 50-60 tokens at largest. My logics in this implementation was the following: most often we'll get around 3 to 4 candidates, and even if we have only 2 penalizable candidates, that would result in almost the same about of iterations going in one loop as going in two loops (find penalizable ones from the front, then go through them and actually penalize). The more penalizable candidates there are, the more effective it is to go in one loop.

Sorting is inevitable here, so I will try to optimize it.

MaggotHATE replied:
I tried replacing sort with a simple shift, but I'm not sure if it's faster (should be, given `std::sort`'s complexity). Additionally, it presents a problem of duplicated tokens at the back if `min_keep` is above 0.

Is it a better solution?

```C++
    int found = 0;
    int pos = 0;
    // going through all candidates from back to front, easier to keep the last of probables
    for (int i = (cur_p->size - 1); i >= 0; --i) {
        if (cur_p->data[i].p >= ctx->threshold && cur_p->data[i].p <= ctx->threshold_max) {
            ++found;
            if (found > 1) {
                // .logits are used for sorting and calculating .p in llama_sample_softmax_impl
                cur_p->data[i].logit = -999.0f;
                pos = i;
            }
        }
    }

    if (found > 1) {
        int removed = found - 1;
        size_t size_new = cur_p->size - removed;

        for (size_t i = pos; i < size_new - pos; ++i) {
            cur_p->data[i] = cur_p->data[i + removed];
        }

        if (size_new < ctx->min_keep) size_new = ctx->min_keep;
        cur_p->size = size_new;
    }
```

MaggotHATE replied:
I think I've found a solution, it should be faster than the previous one. At thin point, however, it might be worth reworking the first for loop.

jukofyork replied:
Looking at:

https://github.com/oobabooga/text-generation-webui/pull/6335

Can't you just use almost the same logic as `min_p`:

```cpp
    // if the cur_p aren't sorted, try the unsorted implementation first
    if (!cur_p->sorted) {
        std::vector<llama_token_data> filtered_tokens;

        float max_logit = -FLT_MAX;
        for (size_t i = 0; i < cur_p->size; ++i) {
            max_logit = std::max(max_logit, cur_p->data[i].logit);
        }
        const float min_logit = max_logit + logf(ctx->p); // min logit for p_i >= p * p_max

        for (size_t i = 0; i < cur_p->size; ++i) {
            if (cur_p->data[i].logit >= min_logit) {
                filtered_tokens.push_back(cur_p->data[i]);
            }
        }

        // if we have enough values the operation was a success
        if (filtered_tokens.size() >= ctx->min_keep) {
            memcpy(cur_p->data, filtered_tokens.data(), filtered_tokens.size()*sizeof(llama_token_data));
            cur_p->size = filtered_tokens.size();
            min_p_applied = true;
        }
    }
 ```

but use `cur_p->data[i].logit <= min_logit` to place into `filtered_tokens`, and then keep track of a single  `best_token` / `best_logit` pair (ie: that is current best that exceeds `min_logit` by the smallest amount). Then at the end prepend `best_logit` before the `memcpy` if found?

Unless I've misunderstood the XTC logic then this seems the simplest way to do it?

MaggotHATE replied:
@jukofyork  While batch-testing in my program I noticed consistently shorter results. Out of worry that it was caused by my current implementation of XTC (it was not, I haven't updated to the latest fixes of the new backend registry) I rewrote it with memcpy similar to what you've proposed:

```C++
    for (int i = 0; i < candidates->size; ++i) {
        if (cur_p->data[i].p < ctx->threshold || cur_p->data[i].p > ctx->threshold_max) {
            if (i > 0 && cur_p->data[i-1].p >= ctx->threshold) cur.emplace_back(cur_p->data[i-1]);
            cur.emplace_back(cur_p->data[i]);
        } else if (pos == -1) pos = i;
    }

    size_t to_remove = cur_p->size - cur.size();

    if (to_remove >= ctx->min && cur.size() >= ctx->min_keep)) {
        memcpy(cur_p->data, cur.data(), cur.size()*sizeof(llama_token_data));
        cur_p->size = cur.size();
    }
```

Is it correct? Am I missing anything?

MaggotHATE committed:
First fixes by comments

Still need to look into sorting

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

MaggotHATE commented on src/llama-sampling.cpp:
```diff
@@ -1059,6 +1059,86 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
+
+    std::random_device rd;
+    float chance = (float)(rd()%100)/100;
+    if (chance > ctx->probability) return;
```
I think I've resolved this now, although I may have misunderstood it. Are chances good now?

slaren replied:
`random_device` should only be used to initialize other RNGs, and we want the results to be repeatable when the seed is the same. You should keep a `std::mt19937` RNG in the sampler state in the same way that other samplers such as `llama_sampler_dist` do it. Take a `seed` parameter and reset the RNG usign this seed in the `reset` function. In this function, use the RNG in the state with `std::uniform_real_distribution` to obtain a number between 0 and 1.

slaren replied:
`reset` is only called when completely resetting the sampler to its original state, so the current implementation will only generate the `chance` once, which is probably not what you want. This is what I meant:

```diff
diff --git a/src/llama-sampling.cpp b/src/llama-sampling.cpp
index 4372b40c..2b368081 100644
--- a/src/llama-sampling.cpp
+++ b/src/llama-sampling.cpp
@@ -1069,9 +1069,8 @@ struct llama_sampler_xtc {

     const uint32_t seed;
     uint32_t       seed_cur;
-    float          chance;

-    std::mt19937 rng;
+    std::mt19937   rng;
 };

 static const char * llama_sampler_xtc_name(const struct llama_sampler * /*smpl*/) {
@@ -1079,7 +1078,7 @@ static const char * llama_sampler_xtc_name(const struct llama_sampler * /*smpl*/
 }

 static void llama_sample_xtc_apply(struct llama_sampler * smpl, llama_token_data_array * cur_p) {
-    const auto * ctx = (llama_sampler_xtc *) smpl->ctx;
+    auto * ctx = (llama_sampler_xtc *) smpl->ctx;

     if (ctx->probability <= 0.0f
         || ctx->threshold <= 0.0f
@@ -1090,8 +1089,10 @@ static void llama_sample_xtc_apply(struct llama_sampler * smpl, llama_token_data
         || ctx->min_keep <= 2) {
         return;
     }
-    // chance is calculated on init and on each reset
-    if (ctx->chance > ctx->probability) return;
+
+    std::uniform_real_distribution<float> distance(0.0f, 1.0f);
+    float chance = distance(ctx->rng);
+    if (chance > ctx->probability) return;

     // in case it's not sorted/recalculated yet
     llama_sampler_softmax_impl(cur_p);
@@ -1141,9 +1142,6 @@ static void llama_sampler_xtc_reset(struct llama_sampler * smpl) {
     auto * ctx = (llama_sampler_xtc *) smpl->ctx;
     ctx->seed_cur = get_rng_seed(ctx->seed);
     ctx->rng.seed(ctx->seed_cur);
-
-    std::uniform_real_distribution<> distance(0.0, 1.0);
-    ctx->chance = distance(ctx->rng);
 }

 static struct llama_sampler_i llama_sampler_xtc_i = {
@@ -1157,9 +1155,6 @@ static struct llama_sampler_i llama_sampler_xtc_i = {

 struct llama_sampler * llama_sampler_init_xtc(float p, float t, float t_max, size_t min_keep, uint32_t seed) {
     auto seed_cur = get_rng_seed(seed);
-    std::uniform_real_distribution<> distance(0.0, 1.0);
-    auto rng = std::mt19937(seed_cur);
-    float chance = distance(rng);
     return new llama_sampler {
         /* .iface = */ &llama_sampler_xtc_i,
         /* .ctx   = */ new llama_sampler_xtc {
@@ -1169,8 +1164,7 @@ struct llama_sampler * llama_sampler_init_xtc(float p, float t, float t_max, siz
             /* .min_keep      = */ min_keep,
             /* .seed          = */ seed,
             /* .seed_cur      = */ seed_cur,
-            /* .chance        = */ chance,
-            /* .rng           = */ rng,
+            /* .rng           = */ std::mt19937(seed_cur),
         },
     };
 }
```

MaggotHATE replied:
Oh, I assumed that `ctx` should stay `const` in `_apply`! Thanks, I definitely misunderstood that.

MaggotHATE committed:
Fixed trailing backspaces

MaggotHATE committed:
Fixed RNG to be reproduceable 

Thanks to @slaren for directions

MaggotHATE committed:
Fixed forgotten header

MaggotHATE committed:
Moved `min_keep` 

Moved from conditions to a simple check at the end.

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

MaggotHATE committed:
Fixed broken randomization

Thanks to @slaren for explanation

David-AU-github commented:
I have tested this with a number of models of different sizes / types via Text Generation / oobabooga on day of release.
This works very well and a strong addition to LLAMACPP.

p-e-w commented:
What is the theory behind the `xtc_t_max` parameter? It seems to me that in many cases, that parameter would do the opposite of what XTC is supposed to do: *Reduce* creativity compared to not using XTC at all.

Here's an example: Imagine the input is `... shiver down his`, and the predictions from the model are

```
spine 0.65
back 0.15
throat 0.12
...
```

With the original XTC algorithm and `xtc_t` set to 0.1, `spine` and `back` will be eliminated, leaving `throat` and the rest of the distribution. The behavior is more creative because the most clichéd continuations are gone.

Now let's say you set `xtc_t` to 0.1 again, and you also set the `xtc_t_max` parameter from this PR to 0.6. In that case, the algorithm will eliminate only `back`, leaving `spine` because it is above the upper threshold. The probability mass of `back` is distributed to the remaining tokens after renormalization, and the most clichéd continuation (`spine`), the one we want to avoid above all else, **is now *even more likely* than it would have been if we had done nothing at all.**

That doesn't seem right.

MaggotHATE commented:
> Here's an example: Imagine the input is `... shiver down his`

These example and explanations are true _for baseline models_, but may not work for finetuned models that were trained against clichéd phrases.

Consider a model which, as a result of finetuning, is now less likely to output the phrase in your example: less likely doesn't mean "erased entirely", and such models may still be able to generate it, but with lower probability. In such case XTC will remove a more probable (supposedly less clichéd) variant and increase the chance of choosing a previously less probable cliché. In the long run it can steer such finetuned model back to its baseline, "undoing" the effect of finetuning to some extent.

At the same time XTC may still be useful for such a model if a middle range is carefully set. I would assume that it is possible to cut out a range where penalized clichés may still appear, further improving model's capabilities. In practice, I have tested only a few such models, but I was getting better (as in, less typical) responses with lowered upper threshold.

You can look at various [benchmarks](https://eqbench.com/creative_writing.html) that estimate creativity and, more relevant in our case, so called "slop", which is exactly the type of phrases we want to avoid using XTC. It's worth testing XTC with such models too.

As a note, while testing new fixed random, I've encountered refusals on `mini-magnum-12b-v1.1.Q8_0` where they never happened even on baseline Nemo Instruct, with parameters `M=0.020->xtc=0.20-0.99 (99% chance)` (and for statistics I was counting tokens that came through XTC sampler and that were removed: `611/37275(1.64%)` ). While this is an extreme example (99% chance), it shows that it is possible to get an undesirable result in some corner cases. The addition of upper threshold simply makes XTC a bit more flexible without breaking it (default value is still 1.0).

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

MaggotHATE committed:
Swapped sorting for a custom algorithm

Shifts tokens to remove the penalized ones, then puts the penalized at the back. Should make `min_keep` still viable.

p-e-w commented:
> Consider a model which, as a result of finetuning, is now less likely to output the phrase in your example: less likely doesn't mean "erased entirely", and such models may still be able to generate it, but with lower probability. In such case XTC will remove a more probable (supposedly less clichéd) variant and increase the chance of choosing a previously less probable cliché. In the long run it can steer such finetuned model back to its baseline, "undoing" the effect of finetuning to some extent.

So the purpose of `xtc_t_max` is to act as a countermeasure so that the finetune's anti-cliché countermeasures don't get overridden by XTC, which is itself an anti-cliché countermeasure? That's highly convoluted.

I suspect that this parameter is useless in most cases, actively harmful unless extremely carefully calibrated for each model and input, and above all else, confusing and very difficult to interpret for most users. The standard XTC threshold gives you a straightforward knob to tweak, with a clearly defined meaning: The probability that you consider "viable". By contrast, with `xtc_t_max`, you have to guess to what degree the finetune has already suppressed clichés, and that degree might not even be uniform across all clichés. Essentially, `xtc_t_max` means nothing at all. It's just a value that you can fiddle with in the hope that you will like the output better.

IMO, `xtc_t_max` is a philosophical mismatch for XTC, injecting complexity and volatility into an otherwise simple and straightforward sampler. I considered several additional parameters while designing XTC, but in the end decided against including any of them, because simplicity is king, and interpretability is what makes parameters useful in the first place.

MaggotHATE commented:
> So the purpose of `xtc_t_max` is to act as a countermeasure so that the finetune's anti-cliché countermeasures don't get overridden by XTC, which is itself an anti-cliché countermeasure?

No, it a measurement of a model's _perceived creativity_. If a model is already creative enough, more clichés will be less probable - and the range of tokens we perceive as undesirable would have lower ceiling.

Look at it from a broader perspective: every model needs a direction to create an output a user might need. To achieve that, we can perform an **internal** work and an **external** one on that model.

**Internal** work is training/finetuning - if affects the internal data of the model, and the results of such work are "baked" into the model itself. While not usually used, it is possible for any well-made model to output coherently (almost) without samplers.

**External** work are sampling algorithms, RAG, etc. They are used to redirect the model towards a more desirable output for every input a user might have. We can also say that prompt engineering falls into this category (as opposed to creating training data pairs, for example, which would be internal work).

Now, let's look at XTC again: the initial idea was to remove the most "boring", repetitive tokens from consideration. This means that we see the results of the internal work done on the model as "incomplete" from what we see as "creativity" of a model. To make it "complete", we use XTC and get exactly (or almost exactly) what we need.

However, if the internal work done one a model has already considered that creativity might be "incomplete", and further training was performed, that model by itself will be closer to what we want to achieve. We see the base output of such model as "more creative" in comparison to another (usually, the base/instruct versions) model. This also means that undesirable tokens will have lower probabilities during sampling.

With the original XTC, however, we will still penalize _top tokens only_. There are no mechanisms to control it's behavior towards upper top tokens, because raising the bar affects lower top tokens, and changing the probabilities affects all top tokens. This means that we don't differentiate between models that are more creative and models that are less creative - lower threshold cannot reflect that as it controls lower top tokens, not the upper ones.

By introducing `xtc_t_max` we give a user more control over the _perceived creativity_, making it easier to find per-model settings. It has always been the case with presets which might be sufficient for one model, but not sufficient for another. Considering the large number of models nowadays, `xtc_t_max` is an important addition and a useful flexibility.

> because simplicity is king

I would argue that functionality is the king: giving users more options and more control over the tools they use is more important than keeping algorithms simple. KISS is important, but it should not hurt usability. In this case: original XTC is a sampler that works with ranges of tokens from top only, with `xtc_t_max` it works with any range of tokens, from top or in the middle. Users who don't need to fiddle with it, will keep it on default, while those who want to experiment will be able to try middle ranges. Possibilities are doubled.

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

MaggotHATE committed:
Algorithm rework

1. Scan token from top till the first non-penalizable
2. Remove the last captured token (the least probable above threshold)
3. Shift all tokens to override the remaining penalizable
4. Penalize and put them at the the bottom.

MaggotHATE commented:
I have reworked the algorithm, given that std::sort is no longer used. It should be ready now.

MaggotHATE commented:
After tinkering with the algorithm more I've tested two options:

The more elegant one is to iterate through all tokens from back to front, which makes capturing the last op token easier:
```C++
    std::vector<llama_token_data> tokens_left;
    bool last_captured = false;

    for (int i = (cur_p->size - 1); i >= 0; --i) {
        if (cur_p->data[i].p < ctx->threshold || cur_p->data[i].p > ctx->threshold_max) {
            tokens_left.emplace_back(cur_p->data[i]);
        } else if (!last_captured) {
            tokens_left.emplace_back(cur_p->data[i]);
            last_captured = true;
        }
    }

    if (last_captured) {
        int to_remove = cur_p->size - tokens_left.size();

        if (to_remove >= 1 && tokens_left.size() >= ctx->min_keep) {
            memcpy(cur_p->data, tokens_left.data(), tokens_left.size()*sizeof(llama_token_data));

            cur_p->size = tokens_left.size();
            // since tokens are in reverse order, we'll need to sort them later
            cur_p->sorted = false;
        }
    }
```
The problem is, it requires sorting later, which will be performed by `softmax` later. That is expensive, and we can avoid it by going from the front:

```C++
    std::vector<llama_token_data> tokens_left;
    bool last_captured = true;

    for (size_t i = 0; i < cur_p->size; ++i) {
        if (cur_p->data[i].p < ctx->threshold || cur_p->data[i].p > ctx->threshold_max) {
            // capturing the last top token
            if (!last_captured && i > 0 && cur_p->data[i-1].p >= ctx->threshold) {
                tokens_left.emplace_back(cur_p->data[i-1]);
                last_captured = true;
            }

            tokens_left.emplace_back(cur_p->data[i]);
        }
    }

    // in case all candidates are penalizable
    if (cur_p->data[cur_p->size - 1].p > ctx->threshold) tokens_left.emplace_back(cur_p->data[cur_p->size - 1]);

    if (tokens_left.size() < ctx->min_keep) return;

    size_t to_remove = cur_p->size - tokens_left.size();

    if (to_remove >= 1) {
        memcpy(cur_p->data, tokens_left.data(), tokens_left.size()*sizeof(llama_token_data));
        cur_p->size = tokens_left.size();
    }
```

However, the second one feels bulky. Which one is more acceptable? @slaren can I request your help here?

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

MaggotHATE commented:
> you should only need to first find the index of the last token to remove, and then move the remaining tokens to the top of the list

It will be slightly more complex due to addition of max threshold, but yes, I understand. I have also added XTC into `test-sampling` for self-check.

> disable itself if any special tokens would be removed

It's not a part of the original idea, and I haven't encountered any problems with special tokens _in llama.cpp specifically_ - that's why I left it out for now. If any problems are found later, I'll definitely add it.

pnb commented:
> ... The implementation in textgen-webui also seems to completely disable itself if any special tokens would be removed, which I imagine is important.

Wouldn't doing so bias the responses to be shorter, on average? If EOT is a special token and can be elevated in probability by XTC when other (more likely) tokens are removed, but never decreased in probability by itself being removed, it seems like responses will be shorter.

I've tried this implementation a bit now and it does occasionally ramble a bit at the end of a response, especially when generating code (not a great use for XTC but entertaining at least). That might be caused by removing EOT, so perhaps there's good reason to avoid it. But I would definitely prefer an occasional ramble over shortened responses.

MaggotHATE commented:
> shortened responses

Interestingly, XTC can behave differently in that regard: for example, `mini-magnum-12b-v1.1.Q8_0` started generating shorter responses with XTC cutting out _middle_ ranges (0.05-0.50), while no such effect happened at full range (default 0.1 - 1.0).

At the same time, other sampling techniques can shorten prompts too - for example, [noise](https://github.com/kalomaze/koboldcpp/releases/tag/noisy-v1) as @kalomaze once implemented. When used with high range it can easily shorten responses in half (but also improve creativity...quite extremely). As such, this effect is not exclusive to XTC.

pnb commented:
Interesting! I guess that makes sense for a model where EOT has a bimodal probability distribution, either quite likely or unlikely, and you cut out the middle part of the distribution where it is less commonly occurring.

MaggotHATE committed:
Added XTC to `test-sampling`

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

LostRuins commented:
I agree with p-e-w, and their original design is what kobold uses.

p-e-w commented:
@MaggotHATE 

> No, it a measurement of a model's *perceived creativity.* If a model is already creative enough, more clichés will be less probable - and the range of tokens we perceive as undesirable would have lower ceiling.

No, that interpretation doesn't work, because the model's innate creativity is context-dependent. It's impossible to say something like "this model places clichéd tokens below 60% probability, so we set `xtc_t_max` to 0.6". Such trained cliché-avoidance behavior will never be uniform. On the other hand, it's perfectly reasonable to say "any token that the model assigns a probability of 10% or greater is likely to make sense, so we set `xtc_t` to 0.1".

The `xtc_t_max` parameter has no plain interpretation, and getting the value wrong will *reinforce* clichéd behavior by shifting probability mass to the top token, something that can never happen with the original XTC parameters, no matter which values you choose. That's the main reason I consider `xtc_t_max` to be bad: If you pick the wrong value, XTC will suddenly do the *opposite* of what it is designed to do. In other words, it's a classic footgun. And no value below 1 is safe a priori; you always have to experiment, and a value so obtained doesn't carry over to any other model or input.

`xtc_t_max` is a hack that may allow certain individual users to adapt certain models to work slightly better than they would otherwise. It can also wreck XTC, and even invert its purpose, if set to a slightly wrong value, or if left unchanged from a previously good value after switching to a different model. It's unstable and unpredictable, and can end up doing the opposite of what the user wants. Such a parameter should not be exposed by software used by so many people, and no other implementation of XTC provides it. If it works well for you, then I recommend keeping it in a fork.

MaggotHATE commented:
@p-e-w 

>that interpretation doesn't work, because the model's innate creativity is context-dependent.

That context is built not only by the user, but also by the model itself: every previous token affects every next token, and that affects the whole result, making control over top tokens more important and nuanced (and is also why XTC is so important in general). If we only control the lower threshold, we wouldn't differentiate between "a baseline model used for creativity" and "a finetuned model used for creativity" - in both cases we would only assume that a baseline model needs a wider range than a finetuned model. That range would be always pinned to the top, which works for baseline, but may not for a finetune. The upper threshold, as an _optional_ parameter, fixes that when applied in the latter case.

> It's impossible to say something like "this model places clichéd tokens below 60% probability, so we set `xtc_t_max` to 0.6". 

> On the other hand, it's perfectly reasonable to say "any token that the model assigns a probability of 10% or greater is likely to make sense, so we set `xtc_t` to 0.1".

I see inconsistency here: we can equally say "most tokens that the model assigns a probability 60% or greater are likely not lead to cliched phrasing, so we set `xtc_t_max` to 0.6" as a **perceived** metric. It's neither precise nor guaranteed, but can help to achieve the main goal - penalizing tokens that lead to cliched phrases.

However, XTC doesn't recognize tokens (unlike [some others](https://github.com/sam-paech/antislop-sampler/)) and can still penalize anything regardless of its value. All we have is a lower limit of token's probability, which makes XTC one-sided. By adding `xtc_t_max` we give it an additional side - it's not a hack, it's a fine-grained control. Without it XTC will be almost useless for a big number of models that were already finetuned against clichés.

> The xtc_t_max parameter has no plain interpretation, and getting the value wrong will reinforce clichéd behavior by shifting probability mass to the top token

This assumes that top tokens always lead to clichés, which would be less true for finetuned models. Again, all these reasons are based on the normal state of things where top tokens are more likely to bring undesirable results. However, there are more  finetuned models than baseline models, and `xtc_t_max` makes XTC more universal without breaking it. A combination of original XTC with some other samplers would not achieve a similar effect since XTC is the only one that controls from top to bottom (well, unless we start restricting tokens directly).

> you always have to experiment, and a value so obtained doesn't carry over to any other model or input.

> In other words, it's a classic footgun

which is why `xtc_t_max` is optional. It doesn't make XTC worse or function differently on default settings. Experimentation is a big part of LLMs even today, and having a more controllable version of XTC would allow that - again, without breaking anything on default settings.

> Such a parameter should not be exposed by software used by so many people

A fair warning in README would be enough, I believe. XTC is turned off from the sampling queue, so users will have to look it up anyway. 

> and no other implementation of XTC provides it.

The same can be said about oobabooga's special tokens handling, and it's more intrusive that `xtc_t_max`. In the end, `xtc_t_max` doesn't break anything, doesn't change what XTC is about and doesn't encourage using it if user don't need it.

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

MaggotHATE committed:
Simplified algorithm and more tests

MaggotHATE commented:
The algorithm is simplified and passes tests. It works slightly faster than memcpy versions according to benchmark. Should be good now.

MaggotHATE committed:
Updated info in common and args

MaggotHATE committed:
Merged back lost commits in common and arg

MaggotHATE committed:
Update dump info in common

slaren commented on src/llama-sampling.cpp:
```diff
@@ -1059,6 +1059,114 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
+        }
+    }
+
+    size_t to_remove = pos_last - (1 + pos_first);
+
+    if (to_remove < ctx->min_keep || to_remove < 1) return;
```
This doesn't look like a correct way to respect `min_keep`. You need to clamp `to_remove` such that at the end `cur_p->size >= min_keep`.

MaggotHATE replied:
Yes, sorry, forgot to change this check.

Although, should I really clamp it or just check for `cur_p->size - to_remove` instead here? Clamping `to_remove` to a specific size sounds strange here. Min_p, for example, just doesn't apply its effect if `min_keep` requirement is not met.

MaggotHATE replied:
On that note, when is `min_keep` even used? It seems counterintuitive for many samplers, yet all of them follow it.

slaren replied:
It's something that has been there from the beginning, I am not sure that it is very useful.

MaggotHATE committed:
Fixed incorrect min_keep check

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

MaggotHATE committed:
Added XTC to README

MaggotHATE commented:
@p-e-w Can you please take a look at the README part about XTC? I tried to to keep it short like with other samplers, but maybe there's something I didn't communicate properly (or maybe recommendations have changed since the original PR).

p-e-w commented on examples/main/README.md:
```diff
@@ -241,6 +241,22 @@ The `--mirostat-ent` option sets the Mirostat target entropy (tau), which repres
+
+-   `--xtc-p N`: Sets the chance for token removal (checked once on sampler start) (default: 0.5).
+-   `--xtc-t N`: Sets a minimum probability threshold for tokens to be removed (default: 0.1).
+-   `--xtc-t-max N`: Sets a maximum probability threshold for tokens to be removed (highly expetrimental) (default: 1.0).
```
```suggestion
-   `--xtc-t-max N`: Sets a maximum probability threshold for tokens to be removed (highly experimental) (default: 1.0).
```

MaggotHATE replied:
In regards to sampling queue: XTC should be the last, right? I ask because it makes sense to have temperature both before and after XTC, but it can change the result significantly - especially with low temperature values (like recommended 0.3 for Nemo). I feel like XTC should go before temperature (bur after other truncating samplers, of course) because XTC is a truncating sampler too.

p-e-w commented on examples/main/README.md:
```diff
@@ -241,6 +241,22 @@ The `--mirostat-ent` option sets the Mirostat target entropy (tau), which repres
+-   `--xtc-t N`: Sets a minimum probability threshold for tokens to be removed (default: 0.1).
+-   `--xtc-t-max N`: Sets a maximum probability threshold for tokens to be removed (highly expetrimental) (default: 1.0).
+
+Exclude Top Choices (XTC) is a unique sampler that is designed to remove top tokens from consideration and avoid more obvious and repetitive answers. With a chance of `xtc-p` it searches for tokens with probabilities of `xtc-t` threshold and above, then removes all such tokens except the least probable one.
```
```suggestion
Exclude Top Choices (XTC) is a unique sampler that is designed to remove top tokens from consideration and avoid more obvious and repetitive outputs. With a chance of `xtc-p` it searches for tokens with probabilities of `xtc-t` threshold and above, then removes all such tokens except the least probable one.
```

p-e-w commented on examples/main/README.md:
```diff
@@ -241,6 +241,22 @@ The `--mirostat-ent` option sets the Mirostat target entropy (tau), which repres
+
+Exclude Top Choices (XTC) is a unique sampler that is designed to remove top tokens from consideration and avoid more obvious and repetitive answers. With a chance of `xtc-p` it searches for tokens with probabilities of `xtc-t` threshold and above, then removes all such tokens except the least probable one.
+
+By removing top tokens XTC can improve variety of answers, break writing clichés and inhibit repition, since clichés and repeated phrases are usually more likely to appear. By keeping the last top token XTC ensures that the answer is still coherent. XTC is meant to be used for creative tasks, but feel free to experiment with different settings for different models.
```
```suggestion
By removing top tokens XTC can improve the variety of answers, break writing clichés and inhibit repition, since clichés and repeated phrases are usually more likely to appear. By keeping the last token above the threshold, XTC ensures that the answer is still coherent. XTC is meant to be used for creative tasks, but feel free to experiment with different settings for different models.
```

p-e-w commented on src/llama-sampling.cpp:
```diff
@@ -1059,6 +1059,114 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
+    auto * ctx = (llama_sampler_xtc *) smpl->ctx;
+
+    if (ctx->probability <= 0.0f
+        || ctx->threshold >= 1.0f
```
```suggestion
        || ctx->threshold > 0.5f
```

Any threshold > 0.5 already disables XTC, since it is impossible for more than one token to have a probability greater than 1/2.

p-e-w reviewed (changes requested):
Please add support for using XTC via the API (`server`), as otherwise most people won't be able to benefit from it.

Make sure to match the semantics of the other API implementations (TGWUI, ExLlamaV2, Kobold):

* The full parameters names `xtc_probability`, `xtc_threshold` must be used instead of the abbreviations from this PR.
* XTC should be in the sampler chain by default, with `xtc_probability` set to 0, so that it can be enabled by setting `xtc_probability` to something > 0, as in the other implementations.

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

MaggotHATE committed:
Renamed parameters, fixed info and defaults

* probability is at 0 by default, but XTC is included in sampling queue
* threshold higher than 0.5 switches XTC off

MaggotHATE committed:
Initial server support

MaggotHATE commented:
I am slightly confused by new server, it's been a while since I used it. Should I change anything else to bring XTC into server API?

ggerganov commented:
It should already work with the current changes. It can be enabled from the client by passing the `"samplers": [..., 'xtc', ...]` in the request.

MaggotHATE commented:
@ggerganov Thank you! The reason I ask is because I started adding XTC to server UI's, but none of two seem to have a correct Mistral template, making it hard to test. But looks like a simple format works well enough.

MaggotHATE committed:
Added XTC to server UIs

MaggotHATE committed:
Merge branch 'master' into master

MaggotHATE committed:
Fixed labels in old server UI

slaren commented:
> Make sure to match the semantics of the other API implementations (TGWUI, ExLlamaV2, Kobold):

Should that include the handling of special tokens? There are some differences in this implementation in that regard.

MaggotHATE commented:
> the handling of special tokens

I'm still testing the server, but so far I have not seen a single problem with special tokens. It should be good as is, both on local (main) and on server.

slaren commented:
Even if there are no apparent problems, users will (rightfully) expect the same behavior between the different implementation of this sampler.

MaggotHATE commented:
This implementation is already different in the upper threshold parameter, so there's no need to conform to existing implementations that closely. Plus, if I remember correctly, Kobold.cpp doesn't have special tokens handling in XTC either.

slaren commented:
I don't think it is strictly necessary to clone the exact same behavior as the original implementation, the idea can be improved, but in that case out of respect for the original author it may be better to use a different name for the sampler. Unless that's ok with @p-e-w.

MaggotHATE committed:
Made algorithm safer and more readable

LostRuins commented:
> This implementation is already different in the upper threshold parameter, so there's no need to conform to existing implementations that closely. Plus, if I remember correctly, Kobold.cpp doesn't have special tokens handling in XTC either.

That's correct, in Kobold there are only two user configurable parameters `xtc_threshold` and `xtc_probability`, which matches p-e-w's original implementation. We did a few tests comparing excluding newlines and found little to no benefit, so the simple approach was taken. 

Further reading here: https://github.com/oobabooga/text-generation-webui/pull/6335#issuecomment-2309330191

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

p-e-w commented:
@slaren 

> I don't think it is strictly necessary to clone the exact same behavior as the original implementation, the idea can be improved, but in that case out of respect for the original author and to avoid confusion, it may be better to use a different name for the sampler. Unless that's ok with @p-e-w.

As I have stated above, the plain implementation from this PR (without special treatment for `\n` and `EOS`) **is** the original design. The special tokens hack was added by the maintainer against my objections because some users were complaining loudly about the behavior with certain large finetunes. As far as I'm concerned, that hack doesn't belong in XTC, and the other implementations (Kobold and ExLlamaV2) thankfully don't have it.

What is **not** part of the original XTC design is the `xtc_t_max` parameter introduced by this PR. No other implementation has that parameter, and as I have explained in detail above, I consider adding that parameter a bad idea because it complicates the user interface with an extremely capricious knob that can actually do the opposite of what XTC is designed for.

MaggotHATE commented:
> What is **not** part of the original XTC design is the `xtc_t_max` parameter introduced by this PR. No other implementation has that parameter, and as I have explained in detail above, I consider adding that parameter a bad idea because it complicates the user interface with an extremely capricious knob that can actually do the opposite of what XTC is designed for.

If a frontend doesn't implement `xtc_t_max` in its UI, nothing will break, though. XTC will work exactly as the original if the upper threshold is left untouched.

p-e-w commented:
If frontends don't expose `xtc_t_max` then the parameter is useless. If frontends do expose it, then users will play with it, the output will become less creative rather than more creative, and people will conclude "XTC doesn't work".

I maintain that `xtc_t_max` doesn't belong in XTC. If you believe otherwise, there is a very simple way to resolve this: Submit that parameter in a separate PR, so that it may be evaluated *on its own merit.*

But please don't use the popularity that XTC has gained to try to shoehorn your own ideas into the sampler by coupling the two, and then submitting them under the title "add XTC sampler". There is nothing wrong in principle with adding parameters to a sampler, including XTC, but that is not the way to do it.

MaggotHATE committed:
Removed xtc_threshold_max

MaggotHATE committed:
Merge branch 'master' of https://github.com/MaggotHATE/llama.cpp-xtc

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

MaggotHATE commented:
@p-e-w 
> to try to shoehorn your own ideas into the sampler

That was never the intention, I simply wanted it to be more flexible after experiencing more cliched responses and refusals on finetuned models, that is all. But I understand what you mean.

> There is nothing wrong in principle with adding parameters to a sampler, including XTC

On that note, as was testing this XTC implementation in server UI, I looked at candidates and saw no consistency in the result any range has on a finetuned model (still using `mini-magnum` for testing). At times XTC pulls out clichés on default settings, but different clichés appear with different ranges. The "dimly lit" is probably the most elusive one from my current tests. However, tokens that usually lead to cliched phases have lower probabilities most of the time and, what's more important, they are more mixed with the tokens that would give less cliched results.

As such, while I'm still sure that there should be a mechanism for more control, I am not longer sure about the top range being the best of them. Alternatives I tried in the past were:

1. _Penalizing each token above threshold regardless of their number_ (i.e. even if there's only one of them) - this has not caused trouble for me in past, but it may reduce prompt adherence. At the same time, top tokens that might not be penalized originally will be penalized with this setting. It will require higher threshold, but will allow to remove all tokens with probabilities 80% and above, for example, making the effect more consistent. Otherwise we would need a threshold that would also catch tokens of at least 50%. Looking at candidates, I see that in a finetuned model the most cliched results are often caused by tokens of probabilities ~80% and higher.
  
3. _Calculating probability for each token_ - that's a minor change, but it may help keeping some tokens above the threshold that we might want to (again, the hypothetical situation where we want to remove everything above 80%, but don't want to remove anything below that). It's inconsistent, though, so a better solution would be the first one.

What are your thoughts on these? I will implement them in a separate PR since there is a potential for improvements.

MaggotHATE committed:
Fixed arg after update

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

p-e-w commented:
> This can be done by penalizing exactly one top token if threshold is above 0.5, I will add it as a separate PR once this gets merged.

That's an interesting proposal, but it sounds like a completely new sampler, rather than an addition to XTC. I recommend submitting that mechanism as a separate sampler under a new name. It's pretty much the inverse of XTC in the way it removes tokens.

MaggotHATE commented:
> I recommend submitting that mechanism as a separate sampler under a new name.

Yes, I tested more in the meantime and see potential in two ideas:
1. A simple range-based removal would allow penalizing from 1.0 till 0.8, for example, doing exactly that.
2. A top tokens disposition would check if the difference between first two tokens is higher than X (in this case, for example, 0.65) and, if true, remove the top token. This would allow removing only the most "dominating" tokens while keeping those with more variety (for example, in candidates 40%, 40%, 10%,etc).

p-e-w commented on src/llama-sampling.cpp:
```diff
@@ -1059,6 +1059,107 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
+        size_t last_idx = cur_p->size - pos_last;
+
+        for (size_t i = 0; i <= last_idx; ++i) {
+            cur_p->data[i] = cur_p->data[i + pos_last];
```
Can't the whole thing be accomplished by just advancing the `cur_p->data` pointer to skip the elements to be removed? That would avoid the need to shift potentially hundreds of thousands of elements.

MaggotHATE replied:
Yep, it's easier since we always remove the front.

MaggotHATE commented on examples/main/README.md:
```diff
@@ -241,6 +241,19 @@ The `--mirostat-ent` option sets the Mirostat target entropy (tau), which repres
+
+By removing top tokens XTC can improve the variety of answers, break writing clichés and inhibit repition, since clichés and repeated phrases are usually more likely to appear. By keeping the last token above the threshold, XTC ensures that the answer is still coherent. XTC is meant to be used for creative tasks, but feel free to experiment with different settings for different models.
+
+Being experimental and unique, XTC is disabled by default. The recommended combination of samplers is Min-P followed by XTC on its default settings: `--sampling-seq mx --min-p 0.02 -xtc-p 0.5`.
```
@p-e-w  While XTC is included into the sampler queue by default, it is put after all other truncating samplers. As such, the recommended combinations of samplers, as per your words in https://github.com/oobabooga/text-generation-webui/pull/6335 , requires passing samplers chain explicitly.

p-e-w replied:
I don't get it. Seems like the order is correct by default (XTC after truncation, which Min-P is). And all samplers are set to neutral (off) by default, right? So what does `--sampling-seq mx` do that wouldn't happen otherwise?

MaggotHATE replied:
@p-e-w  Not all samplers - Top K is 40 by default and Top P is 0.95. So, either they need to be set to 0 and 1.0 respectively, or (which is easier and more logical) sampling queue should be limited to only the samplers we need.

p-e-w replied:
:open_mouth:  I had no idea! So llama.cpp users are getting crap samplers from the Stone Age without even realizing it. That's terrible. I would have expected a clean slate that samples from the raw model distribution unless parameters are set explicitly.

Anyway, this means your example command is correct, of course.

MaggotHATE replied:
If I remember correctly, this topic has already been brought up some time ago in other PRs. However, since in most cases llama.cpp is used as a library through another app or as a server, this issue is mostly related to llama-cli users. You can look at `index-new.html` (new server UI): it has different "default" values with `top_k` and `top_p` turned off, and I assume any other frontend will have a payload with all parameters set as needed.

But yes, this is an issue.

strawberrymelonpanda replied:
Llama-cli user here via scripts... thanks for bringing this up at any rate, Improved my (custom) benchmark scores just by adjusting settings to disable Top K and Top P.

MaggotHATE commented on src/llama-sampling.cpp:
```diff
@@ -1059,6 +1059,107 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
+    delete (llama_sampler_xtc *) smpl->ctx;
+}
+
+static void llama_sampler_xtc_reset(struct llama_sampler * smpl) {
```
AFAIK this is necessary to properly reset seed and maintain repeatability, as recommended by @slaren earlier.

MaggotHATE commented on src/llama-sampling.cpp:
```diff
@@ -1059,6 +1059,107 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
+    int pos_last = 0;
+
+    for (size_t i = 0; i < cur_p->size; ++i) {
+        if (cur_p->data[i].p - ctx->threshold >= -1e-5) {
```
I added it after running tests - they were failing due to precision problem.

p-e-w replied:
Ugh, no. Never change correct code to satisfy tests. If the code is semantically correct, and the tests don't pass, the tests need to be adapted to account for things such as floating point shenanigans. But changing the code itself is always wrong, unless of course the code has a bug.

The correct way to express this condition is

```c
if (cur_p->data[i].p >= ctx->threshold) {
```

Nothing else will do. And the tests need to work with that, or the tests are wrong.

MaggotHATE replied:
I'm not sure it should be considered a problem with tests (precision problem is a wider topic), but alright.

p-e-w replied:
My basic point is that the tests serve the code, not the other way round. Code is only ever changed in response to failing tests if the code is found to have a bug.

I had a similar problem with tests for an experimental sampler I wrote for llama.cpp a while ago, and I was able to work around it by using a different set of token probabilities in the tests that I constructed specifically so that after probability renormalization, the resulting values were exactly representable in floating point.

MaggotHATE commented on src/llama-sampling.cpp:
```diff
@@ -1059,6 +1059,107 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
+        return;
+    }
+
+    std::uniform_real_distribution<float> distance(0.0f, 1.0f);
```
The name was in code suggested by @slaren earlier. Does it matter here, though? I don't see an issue with such name for a distribution.

slaren replied:
I just moved the declaration, the name was already `distance` in your code. I agree that it is not a very good name.

MaggotHATE commented on common/common.h:
```diff
@@ -108,6 +110,8 @@ struct common_sampler_params {
     float   top_p             = 0.95f; // 1.0 = disabled
     float   min_p             = 0.05f; // 0.0 = disabled
+    float   xtc_probability   = 0.00f; // 0.0 = disabled
+    float   xtc_threshold     = 0.10f; // 0.5 = disabled
```
@p-e-w I think these comments exist as a warning which values of each parameter disable the corresponding sampler. I'll correct these nonetheless.

MaggotHATE committed:
Quick fixes by comments

MaggotHATE committed:
Merge branch 'master' of https://github.com/MaggotHATE/llama.cpp-xtc

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

MaggotHATE committed:
Simplified algorithm since threshold_max is removed

MaggotHATE committed:
Renamed random distribution

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

p-e-w commented on examples/main/README.md:
```diff
@@ -241,6 +241,19 @@ The `--mirostat-ent` option sets the Mirostat target entropy (tau), which repres
+
+Being experimental and unique, XTC is disabled by default. The recommended combination of samplers is Min-P followed by XTC on its default settings: `--sampling-seq mx --min-p 0.02 -xtc-p 0.5`.
+
+Example usage: `-xtc-p 0.5 -xtc-t 0.1
```
Those argument names are not valid anymore. Same above.

MaggotHATE committed:
Fixed tests and outdated README

MaggotHATE committed:
Merge branch 'master' of https://github.com/MaggotHATE/llama.cpp-xtc

p-e-w commented on examples/main/README.md:
```diff
@@ -241,6 +241,19 @@ The `--mirostat-ent` option sets the Mirostat target entropy (tau), which repres
+
+Being experimental and unique, XTC is disabled by default. The recommended combination of samplers is Min-P followed by XTC on its default settings: `--sampling-seq mx --min-p 0.02 --xtc-probability 0.5`.
+
+Example usage: `--xtc-probability 0.5 --xtc-threshold 0.1
```
```suggestion
Example usage: `--xtc-probability 0.5 --xtc-threshold 0.1`
```

p-e-w commented on src/llama-sampling.cpp:
```diff
@@ -1059,6 +1059,101 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
+
+    if (cur_p->size - pos_last >= ctx->min_keep && pos_last > 0) {
+        cur_p->data += pos_last;
+        cur_p->size = cur_p->size - pos_last;
```
```suggestion
        cur_p->size -= pos_last;
```

p-e-w reviewed (approved):
I did another full review of the code and while some parts are quite subtle, I'm now pretty confident that the implementation is correct (though I did not run it myself yet).

@slaren: This is ready to go as far as I'm concerned.

MaggotHATE committed:
Small fixes

MaggotHATE committed:
Merge branch 'ggerganov:master' into master

slaren commented on src/llama-sampling.cpp:
```diff
@@ -1059,6 +1059,101 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
+    }
+
+    if (cur_p->size - pos_last >= ctx->min_keep && pos_last > 0) {
+        cur_p->data += pos_last;
```
This may potentially break 3rd party code that expects this pointer to be unchanged (eg. to free it after sampling). I don't think this is necessarily a problem, but we should make it clear that this pointer may be changed by the samplers, and applications should not rely on it being unchanged.
